docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,833 @@
+key: this subdirectory is the file system of that channel. Within this file system, a
+`meta.json` file is used to store the metadata of this channel, e.g. key, rate, etc.
+This file is not relevant for our purposes of analyzing race conditions.
+
+The channel's file system also includes the `index.domain` file used for domain
+index persist.
+All other files in the file system are used to store binary telemetry data – they are
+numerically labeled and more files get created as existing files get filled up – i.e.
+`1.domain`, `2.domain`, and so on.
+
+<div align="center">
+<img src="img/0019-240529-cesium-race-conditions/File-System-Structure.png" width="90%">
+<h6>File-System structure</h6>
+</div>
+
+### 4.0.2 Cesium's database structure
+
+At the highest level of abstraction, Cesium provides one unified interface to read, write,
+delete, and stream data from multiple channels.
+A Cesium DB comprises many channels, each one of which corresponding to a virtualDB
+or an unaryDB. A virtualDB contains no data, so it will not be a topic of our discussion.
+
+A unaryDB is a second level of abstraction in the database structure: it is composed of
+a domainDB an index, and the controller:
+the index is responsible for querying the timestamps corresponding to telemetry data,
+the controller is responsible for controlling writes to the database between different
+writers, and the domainDB is responsible for storing the actual telemetry data.
+
+A domainDB is the lowest level of abstraction: it provides means to write and read data from
+the file system, and maintains the domain index.
+
+<div align="center">
+<img src="img/0019-240529-cesium-race-conditions/Cesium-Structure.png" width="90%">
+<h6>Database structure</h6>
+</div>
+
+Each layer has their own entities and methods of reads, writes, deletions, and Garbage Collection;
+all interactions are passed down to the domain level to interact with the index and file system.
+
+## 4.1 Top-down view of race conditions in Cesium
+
+### 4.1.0 Types of possible race conditions
+
+There are three types of race conditions that we are most concerned with. The first is
+concurrent modification of the same resource: this must be strictly forbidden as it is
+almost always erroneous. Consider the following code that intends to add 3 and subtract
+1 from 1 – as we know, this should result in 3.
+
+```go
+var sharedInt = 1
+
+// goroutine 1
+func add(){
+    sharedInt += 3
+}
+
+// goroutine 2
+func sub(){
+    sharedInt -= 1
+}
+
+go add()
+go sub()
+```
+
+Both goroutines try to access the same variable, `sharedInt`, leading to potential errors:
+for example, if `add` and `sub` are run at the same time, they may each read `x`'s value
+as 1 and effectuate their operation on 1, leading the end result to become 0 or 4 as
+opposed to the desired value of 3.
+
+The above type of race conditions can be caught thanks to golang's built-in race detector.
```

Comment:
Move this sentence to the above paragraph so each paragraph is only talking about one of the three types of race conditions.